awx/api/views/__init__.py

```diff
@@ -3870,13 +3870,34 @@ def get(self, request, **kwargs):
         # key is counter of meta events (i.e. verbose), value is uuid of the assigned parent
         map_meta_counter_nested_uuid = {}
 
+        # collapsable tree view in the UI only makes sense for tree-like
+        # hierarchy. If ansible is ran with a strategy like free or host_pinned, then
+        # events can be out of sequential order, and no longer follow a tree structure
+        # E1
+        #  E2
+        # E3
+        #  E4  <- parent is E3
+        #  E5  <- parent is E1
+        # in the above, there is no clear way to collapse E1, because E5 comes after
+        # E3, which occurs after E1. Thus the tree view should be disabled.
+
+        # mark the last seen uuid at a given level (0-3)
+        # if a parent uuid is not in this list, then we know the events are not tree-like
+        # and return a response with is_tree: False
+        level_current_uuid = [None, None, None, None]
+
         prev_non_meta_event = events[0]
         for i, e in enumerate(events):
             if not e['event'] in JobJobEventsChildrenSummary.meta_events:
                 prev_non_meta_event = e
             if not e['uuid']:
                 continue
+            level_current_uuid[models.JobEvent.LEVEL_FOR_EVENT[e['event']]] = e['uuid']
```

Comment:
should we ensure lower-level uuids are set to None here? e.g. if the current event's level is 1, then`level_current_uuid[2] = None` and `level_current_uuid[3] = None` to prevent uuids of prior leaf nodes from lingering in the array

Reply:
good point, changed